docs(dynamic-plugins): update and complete the docs (#752)
janus-idp/backstage-showcase: showcase-docs/dynamic-plugins.md

# Dynamic Plugins support

## Overview

This document describes how to enable the dynamic plugins feature in the Janus Backstage showcase application.

## How it works

The dynamic plugin support is based on the [backend plugin manager package](https://github.com/backstage/backstage/tree/master/packages/backend-plugin-manager), which is a service that scans a configured root directory (`dynamicPlugins.rootDirectory` in the app config) for dynamic plugin packages, and loads them dynamically.

> While this package remains in an experimental phase and is held privately in the upstream backstage repository, it is primarily awaiting seamless integration with the new backend system before its APIs can be finalized and frozen. It is worth noting that it is already in use in the backstage showcase application, facilitated by a derivative package published in the `@janus-idp` NPM organization.

Reviewer: ```suggestion
While this package remains in an experimental phase and is a private package in the upstream backstage repository, it is primarily awaiting seamless integration with the new backend system before its APIs can be finalized and frozen. It is worth noting that it is already in use in the backstage showcase application, facilitated by a derivative package published in the `@janus-idp` NPM organization.
```

So this documentation, in its current form, is provisional and will be updated as next PRs related to this feature are merged.

## Overview

This document describes how to enable the dynamic plugins feature in the Janus Backstage showcase application.

## How it works

The dynamic plugin support is based on the [backend plugin manager package](https://github.com/backstage/backstage/tree/master/packages/backend-plugin-manager), which is a service that scans a configured root directory (`dynamicPlugins.rootDirectory` in the app config) for dynamic plugin packages, and loads them dynamically.

> This package is still in experimental status and private in the upstream backstage repository,
> mostly waiting for a complete integration with the new backend system before finalizing and freezing its APIs.
> 
> However it is already used in the backstage showcase application, through a derivative package published in the `@janus-idp` NPM organization.

Reviewer: ```suggestion
While this package remains in an experimental phase and is held privately in the upstream backstage repository, it is primarily awaiting seamless integration with the new backend system before its APIs can be finalized and frozen. It is worth noting that it is already in use in the backstage showcase application, facilitated by a derivative package published in the `@janus-idp` NPM organization.
```

export * from './dynamic/index';
```

#### Note about the new backend system support

> The new backend system is not yet used in the Showcase application.
> However, the underlying mechanism that discovers and loads dynamic backend plugins already supports both the new and old backend systems.
> 
> When the new backend system becomes generally adopted, and used in the janus Showcase, the dynamic backend plugins should then be based on the new backend system.
> 
> Therefore, we **strongly recommend** the creation of the expected new backend system entrypoints (with `createBackendPlugin` or `createBackendModule`) when implementing code changes to make a backend plugin dynamic, in order to be prepared for the move to the new backend system.
> 
> You can find an example of such a dynamic plugin entrypoint based on the new backend system in the [Janus plugins repository](https://github.com/janus-idp/backstage-plugins/blob/main/plugins/aap-backend/src/dynamic/alpha.ts#L14).

Reviewer: ```suggestion
While the Showcase application has not yet integrated the new backend system, the underlying mechanism responsible for discovering and loading dynamic backend plugins is already compatible with both the new and old backend systems.

As the new backend system gains broader acceptance and is implemented in the Janus Showcase, it is advisable to adapt dynamic backend plugins to rely on this new system. Therefore, we **strongly recommend** creating the anticipated entry points for the new backend system (using `createBackendPlugin` or `createBackendModule`) when implementing code changes to enable a backend plugin's dynamic functionality. This proactive step ensures preparedness for the eventual transition to the new backend system.

For a practical example of a dynamic plugin entry point built upon the new backend system, please refer to the [Janus plugins repository](https://github.com/janus-idp/backstage-plugins/blob/main/plugins/aap-backend/src/dynamic/alpha.ts#L14).
```

export * from './dynamic/index';
```

#### Note about the new backend system support

The new backend system is not yet used in the Showcase application.
However the underlying mechanism that discovers and loads dynamic backend plugins, already supports both the new and old backend systems.

> When the new backend system would be generally adopted, and used in the janus Showcase as well, the dynamic backend plugins should be based on the new backend system.

Reviewer 2: ```suggestion
When the new backend system becomes generally adopted, and used in the janus Showcase, the dynamic backend plugins should then be based on the new backend system 
```

> So when implementing code changes to make a backend plugin dynamic, we **strongly recommend** to create the expected new backend system entrypoints (with `createBackendPlugin` or `createBackendModule`), in order to be prepared for the move to the new backend system.

Reviewer 2: ```suggestion
Therefore, we **strongly recommend** the creation of the expected new backend system entrypoints (with `createBackendPlugin` or `createBackendModule`) when implementing code changes to make a backend plugin dynamic, in order to be prepared for the move to the new backend system.
```

> The new backend system is not yet used in the Showcase application.
> However the underlying mechanism that discovers and loads dynamic backend plugins, already supports both the new and old backend systems.
> 
> When the new backend system would be generally adopted, and used in the janus Showcase as well, the dynamic backend plugins should be based on the new backend system.
> 
> So when implementing code changes to make a backend plugin dynamic, we **strongly recommend** to create the expected new backend system entrypoints (with `createBackendPlugin` or `createBackendModule`), in order to be prepared for the move to the new backend system.
> 
> You can find an example of such a dynamic plugin entrypoint based on the new backend system in the [Janus plugins repository](https://github.com/janus-idp/backstage-plugins/blob/main/plugins/aap-backend/src/dynamic/alpha.ts#L14).

Reviewer: ```suggestion
The Showcase application has not yet integrated the new backend system. However, the underlying mechanism that handles the discovery and loading of dynamic backend plugins already supports both the new and old backend systems.

Once the new backend system is widely adopted and implemented in the Janus Showcase, it is recommended to base dynamic backend plugins on this new system. Therefore, when making code changes to enable a backend plugin to be dynamic, we **strongly recommend** creating the expected entry points for the new backend system (using `createBackendPlugin` or `createBackendModule`). This ensures readiness for the transition to the new backend system.

For an example of such a dynamic plugin entry point based on the new backend system, refer to the [Janus plugins repository](https://github.com/janus-idp/backstage-plugins/blob/main/plugins/aap-backend/src/dynamic/alpha.ts#L14).
```

export * from './dynamic/index';
```

#### Note about the new backend system support

The new backend system is not yet used in the Showcase application.

> However the underlying mechanism that discovers and loads dynamic backend plugins, already supports both the new and old backend systems.

Reviewer 2: ```suggestion
However, the underlying mechanism that discovers and loads dynamic backend plugins already supports both the new and old backend systems.
```

Common backstage dependencies, expected to be in the backstage backend application, are not bundled in the dynamic plugin but rather changed as peer dependencies, so that they can be shared with the backstage backend application.

#### Publishing the dynamic backend plugin package to an NPM registry

> The dynamic plugin package, in the `dist-dynamic` sub-folder of the plugin folder, can be published to an NPM registry, using the basic `npm publish` command (since it is not part of any Yarn monorepo).
> 
> As already mentioned above, this published self-contained package will contain its private dependencies in an embedded `node_modules` folder, and will be ready for installation as a dynamic plugin package in the showcase application, as explained [below](#helm-deployment)

Reviewer: ```suggestion
The dynamic plugin package, located in the `dist-dynamic` sub-folder within the plugin directory, can be uploaded to an NPM registry using the standard `npm publish` command as it is not part of any Yarn monorepo.

As previously mentioned, this independently published package is self-contained and includes its own private dependencies in an enclosed `node_modules` folder. It is then prepared for installation as a dynamic plugin package in the showcase application, as elaborated further in the [Helm Deployment](#helm-deployment) section below.
```

Common backstage dependencies, expected to be in the backstage backend application, are not bundled in the dynamic plugin but rather changed as peer dependencies, so that they can be shared with the backstage backend application.

#### Publishing the dynamic backend plugin package to an NPM registry

The dynamic plugin package, in the `dist-dynamic` sub-folder of the plugin folder, can be published to an NPM registry, using the basic `npm publish` command (since it is not part of any Yarn monorepo).

> As already mentioned above this published self-contained package will contain its private dependencies in an embedded `node_modules`` folder, and will be ready to be installed as a dynamic plugin package in the showcase application, as explained [below](#helm-deployment)

Reviewer 2: ```suggestion
As already mentioned above, this published self-contained package will contain its private dependencies in an embedded `node_modules` folder, and will be ready to be installed as a dynamic plugin package in the showcase application, as explained [below](#helm-deployment)
```

> The dynamic plugin package, in the `dist-dynamic` sub-folder of the plugin folder, can be published to an NPM registry, using the basic `npm publish` command (since it is not part of any Yarn monorepo).
> 
> As already mentioned above this published self-contained package will contain its private dependencies in an embedded `node_modules`` folder, and will be ready to be installed as a dynamic plugin package in the showcase application, as explained [below](#helm-deployment)

Reviewer: ```suggestion
The dynamic plugin package, located in the `dist-dynamic` sub-folder within the plugin directory, can be uploaded to an NPM registry with the standard `npm publish` command, as it is not a component of any Yarn monorepo.

As previously stated, this independently published package is self-contained and encompasses its own private dependencies in an enclosed `node_modules` directory. It is then prepared for installation as a dynamic plugin package within the showcase application, as detailed further in the [Helm Deployment](#helm-deployment) section below.
```

### Helm deployment

- Starting from version 2.10.1, the [helm chart](https://github.com/janus-idp/helm-backstage) for deploying the showcase application introduces new values.

> - The updated Helm values introduce a new global.dynamic parameter, comprising two fields:

Reviewer 3: nit:

```suggestion
- The updated Helm values introduce a new `global.dynamic` parameter, comprising two fields:
```

### Helm deployment

> - Starting from version 2.10.1, the helm chart for deploying the showcase application through [this helm chart](https://github.com/janus-idp/helm-backstage) introduces new values.

Reviewer 2: ```suggestion
- Starting from version 2.10.1, the [helm chart](https://github.com/janus-idp/helm-backstage) for deploying the showcase application introduces new values.
```

> - These new Helm values contain a new `global.dynamic` value, with 2 fields:

Reviewer: ```suggestion
- The updated Helm values introduce a new global.dynamic parameter, comprising two fields:
```

>   -  `plugins`: contains the list of dynamic plugins to be installed, and by default is an empty list. A package can be specified either as a local path to the dynamic plugin `dist-dynamic/dist` sub-folder, or as a package specification in an NPM repository. 

Reviewer: ```suggestion
  -  `plugins`: This field encompasses the list of dynamic plugins slated for installation. By default, it is an empty list. A package can be specified either as a local path to the dynamic plugin's dist-dynamic/dist sub-folder or as a package specification in an NPM repository.
```

>   - `includes`: contains a list of YAML files with the same syntax, of which `plugins` list will be included, and possibly overwritten by the `plugins` list of the main helm values. By default, the `includes` fields contains the [`dynamic-plugins.default.yaml`](https://github.com/janus-idp/backstage-showcase/blob/main/dynamic-plugins.default.yaml) file, which contains all the dynamic plugins [shipped with the showcase application](#dynamic-plugins-included-in-the-showcase-container-image), either enabled or disabled by default.

Reviewer: ```suggestion
  - `includes`: This field contains a roster of YAML files utilizing the same syntax. The `plugins` list from these files will be incorporated, and potentially overridden, by the `plugins` list in the primary helm values. The default configuration includes the [`dynamic-plugins.default.yaml`](https://github.com/janus-idp/backstage-showcase/blob/main/dynamic-plugins.default.yaml) file, which encompasses all the dynamic plugins [included in the showcase application container image](#dynamic-plugins-included-in-the-showcase-container-image), whether enabled or disabled by default.
```

> - Adding a dynamic plugin to the showcase is done by adding an entry to the `global.dynamic.plugins` list. Each entry has the following fields:

Reviewer: ```suggestion
- To include a dynamic plugin in the showcase, you can achieve this by appending an entry to the `global.dynamic.plugins` list. Each entry should contain the following fields:
```

### Helm deployment

> - In order to enable dynamic plugins support in the showcase application deployed through the [helm chart](https://github.com/janus-idp/helm-backstage), the helm chart provides new values since version 2.10.1.

Reviewer: ```suggestion
- Starting from version 2.10.1, the helm chart for deploying the showcase application through [this helm chart](https://github.com/janus-idp/helm-backstage) introduces new values.
```

> - These new Helm values contain a new `global.dynamic` value, with 2 fields: `plugins` and `includes`. `plugins` contains the list of dynamic plugins to be installed, and by default is an empty list. A package can be specified either as a local path to the dynamic plugin `dist-dynamic/dist` sub-folder, or as a package specification in an NPM repository. `includes` contains a list of YAML files with the same syntax, of which `plugins` list will be included, and possibly overwritten by the `plugins` list of the main helm values. By default the `includes` fields contains the [`dynamic-plugins.default.yaml`](https://github.com/janus-idp/backstage-showcase/blob/main/dynamic-plugins.default.yaml) file, which contains all the dynamic plugins [shipped with the showcase application](#dynamic-plugins-included-in-the-showcase-container-image), either enabled or disabled by default.

Reviewer 2: ```suggestion
- These new Helm values contain a new `global.dynamic` value, with 2 fields:

  -  `plugins`: contains the list of dynamic plugins to be installed, and by default is an empty list. A package can be specified either as a local path to the dynamic plugin `dist-dynamic/dist` sub-folder, or as a package specification in an NPM repository. 
  - `includes`: contains a list of YAML files with the same syntax, of which `plugins` list will be included, and possibly overwritten by the `plugins` list of the main helm values. By default, the `includes` fields contains the [`dynamic-plugins.default.yaml`](https://github.com/janus-idp/backstage-showcase/blob/main/dynamic-plugins.default.yaml) file, which contains all the dynamic plugins [shipped with the showcase application](#dynamic-plugins-included-in-the-showcase-container-image), either enabled or disabled by default.

```

> - These new Helm values contain a new `global.dynamic` value, with 2 fields: `plugins` and `includes`. `plugins` contains the list of dynamic plugins to be installed, and by default is an empty list. A package can be specified either as a local path to the dynamic plugin `dist-dynamic/dist` sub-folder, or as a package specification in an NPM repository. `includes` contains a list of YAML files with the same syntax, of which `plugins` list will be included, and possibly overwritten by the `plugins` list of the main helm values. By default the `includes` fields contains the [`dynamic-plugins.default.yaml`](https://github.com/janus-idp/backstage-showcase/blob/main/dynamic-plugins.default.yaml) file, which contains all the dynamic plugins [shipped with the showcase application](#dynamic-plugins-included-in-the-showcase-container-image), either enabled or disabled by default.

Reviewer: ```suggestion
- These updated Helm values feature a `global.dynamic` value with two subfields: `plugins` and `includes`. The `plugins` field comprises a list of dynamic plugins to be installed, initially empty. A package can be specified either as a local path to the dynamic plugin's `dist-dynamic/dist` sub-folder or as a package specification in an NPM repository. The `includes` field contains a list of YAML files with the same syntax, where the `plugins` list will be incorporated and potentially overwritten by the `plugins` list in the main helm values. By default, the `includes` field includes the [`dynamic-plugins.default.yaml`](https://github.com/janus-idp/backstage-showcase/blob/main/dynamic-plugins.default.yaml) file, which encompasses all the dynamic plugins [included in the showcase application container image](#dynamic-plugins-included-in-the-showcase-container-image), whether enabled or disabled by default.
```

> - So adding a dynamic plugin to the showcase is done by adding an entry to the `global.dynamic.plugins` list. Each entry has the following fields:

Reviewer 2: ```suggestion
- Adding a dynamic plugin to the showcase is done by adding an entry to the `global.dynamic.plugins` list. Each entry has the following fields:
```

          disabled: false
  ```

### Dynamic plugins included in the Showcase container image

The showcase docker image contains a set of ready-to-use dynamic plugins, most of which are disabled by default, since they require some mandatory configuration.
The list of those plugins can be found in the [`dynamic-plugins.default.yaml`](https://github.com/janus-idp/backstage-showcase/blob/main/dynamic-plugins.default.yaml) file.

At application start, for every plugin disabled by default, the log of the `install-dynamic-plugins` init container of the `backstage` Pod will contain a line like below:

```
======= Skipping disabled dynamic plugin ./dynamic-plugins/dist/backstage-plugin-catalog-backend-module-github-dynamic
```

Enabling this plugin is very easy: just add a package with the same name and override the `disabled` field in the helm chart values:

```diff
global:
  dynamic:
    includes:
      - dynamic-plugins.default.yaml
    plugins:
+      - package: ./dynamic-plugins/dist/backstage-plugin-catalog-backend-module-github-dynamic
+        disabled: false
```

The plugin default configuration is read from the `dynamic-plugins.default.yaml`, but

> you can still override it by adding a `pluginConfig` entry to the plugin configuration.

Reviewer 2: Just a quick question about this, what is the priority for these configurations? Suppose the end user decides to manually configure these values in their `app-config.yaml` via a separate ConfigMap, will the configurations in the `dynamic-plugins.default.yaml` take precedence? 

Also is there a way to enable the plugin BUT not enable the default configurations? Since there may be some configuration conflicts with the default values if those get loaded. 

For example, the `ocm-backend` plugin allows configurations for the entity provider with `catalog.provider.ocm.<arbitrary-key-name>`, however in the default configurations in [dynamic-plugins.default.yaml](https://github.com/janus-idp/backstage-showcase/blob/3a350189e714f5aa4980c1724539f822abcf33f5/dynamic-plugins.default.yaml#L382-L394) this `arbitrary-key-name` is `default`. 

So if this default configuration is loaded and the user tries to override it with their own configurations like below (with a different `arbitrary-key-name` than the one in our default (`default`)), the backend will fail complaining about missing configurations for `ocm.default`. 
```yaml
catalog:
  providers:
     ocm:
        test:
           # other configurations
```

This was the behavior when we had the plugin installed statically and had the "default" config keys defined in the `app-config.yaml` (now [`app-config.deprecated.yaml`](https://github.com/janus-idp/backstage-showcase/blob/3a350189e714f5aa4980c1724539f822abcf33f5/app-config.deprecated.yaml#L233-L238))

Author: When I write:

> you can still override it by adding a `pluginConfig` entry to the plugin configuration.

I'm mean:

> replace the whole default configuration fragment by the one you will provide in the `pluginConfig` field.

I hope this answers your last question.

Reviewer 2: Ok then this shouldn't be an issue

> The showcase docker image contains a set of ready-to-use dynamic plugins, most of which are disabled by default, since they require some mandatory configuration.
> The list of those plugins can be found in the [`dynamic-plugins.default.yaml`](https://github.com/janus-idp/backstage-showcase/blob/main/dynamic-plugins.default.yaml) file.
> 
> At application start, for every plugin disabled by default, the log of the `install-dynamic-plugins` init container of the `backstage` Pod will contain a line like below:
> 
> ```
> ======= Skipping disabled dynamic plugin ./dynamic-plugins/dist/backstage-plugin-catalog-backend-module-github-dynamic
> ```
> 
> Enabling this plugin is very easy: just add a package with the same name and override the `disabled` field in the helm chart values:
> 
> ```diff
> global:
>   dynamic:
>     includes:
>       - dynamic-plugins.default.yaml
>     plugins:
> +      - package: ./dynamic-plugins/dist/backstage-plugin-catalog-backend-module-github-dynamic
> +        disabled: false
> ```
> 
> The plugin default configuration is read from the `dynamic-plugins.default.yaml`, but
> you can still override it by adding a `pluginConfig` entry to the plugin configuration.

Reviewer: ````suggestion
The showcase docker image comes pre-loaded with a selection of dynamic plugins, with most being initially deactivated due to the need for mandatory configuration. The complete list of these plugins is detailed in the [`dynamic-plugins.default.yaml`](https://github.com/janus-idp/backstage-showcase/blob/main/dynamic-plugins.default.yaml) file.

Upon application startup, for each plugin that is disabled by default, the `install-dynamic-plugins` init container within the `backstage` Pod's log will display a line similar to the following:

```
======= Skipping disabled dynamic plugin ./dynamic-plugins/dist/backstage-plugin-catalog-backend-module-github-dynamic
```

To activate this plugin, simply add a package with the same name and adjust the `disabled` field in the helm chart values as shown below:

```diff
global:
  dynamic:
    includes:
      - dynamic-plugins.default.yaml
    plugins:
+      - package: ./dynamic-plugins/dist/backstage-plugin-catalog-backend-module-github-dynamic
+        disabled: false
```

While the plugin's default configuration is extracted from the `dynamic-plugins.default.yaml` file, you still have the option to override it by incorporating a `pluginConfig` entry into the plugin configuration.
````

### Example of external dynamic backend plugins

> If you want to easily test installing dynamic backend plugins from a remote NPM registry,

Reviewer: ```suggestion
If you want to easily test the installation of dynamic backend plugins from a remote NPM registry,
```
